enterprise/backend/src/metabase_enterprise/task/cache.clj

```diff
@@ -4,23 +4,237 @@
+                         [:= :qe.error nil]
+                         [:= :qe.is_sandboxed false]
+                         (if parameterized?
+                           [:and
+                            [:= :qe.parameterized true]
+                             ;; Only rerun a parameterized query if it's had a cache hit within the last caching window
+                            [:= :qe.cache_hit true]
+                             ;; Don't factor the last cache refresh into whether we should rerun a parameterized query
+                            [:not= :qe.context (name :cache-refresh)]]
+                           [:= :qe.parameterized false])]
+              :group-by [:q.query_hash :q.query :qe.card_id]}}))]
+    {:select [:u.query :u.card-id :u.count]
+     :from   [[{:union queries} :u]]}))
+
+(defn- cache-configs->card-ids
+  "Takes a list of cache configs and returns a map from cache configs to relevant card IDs. For cache configs defined on
+  a dashboard, this includes all cards on the dashboard."
+  [cache-configs]
+  (let [dashboard-ids          (->> (filter #(= (:model %) "dashboard") cache-configs)
+                                    (map :model_id))
+        dashboards             (when (seq dashboard-ids)
+                                 (-> (t2/select :model/Dashboard {:where [:in :id dashboard-ids]})
+                                     (t2/hydrate :dashcards)))
+        dashboard-id->card-ids (into {}
+                                     (map (fn [dashboard] [(:id dashboard) (mapv :card_id (:dashcards dashboard))])
+                                          dashboards))
+        config-to-card-ids     (map (fn [config]
+                                      (let [model (:model config)
+                                            model-id (:model_id config)]
+                                        [config
+                                         (cond
+                                           (= model "dashboard") (get dashboard-id->card-ids model-id [])
+                                           (= model "question") [model-id]
+                                           :else [])]))
+                                    cache-configs)]
+    (into {} config-to-card-ids)))
+
+(defn- select-parameterized-queries
+  [parameterized-queries]
+  (apply concat
+         (-> (group-by :card-id parameterized-queries)
+             (update-vals
+              (fn [queries]
+                (->> queries
+                     (sort-by :count >)
+                     (take *parameterized-queries-to-rerun-per-card*))))
+             vals)))
+
+(defn- duration-queries-to-rerun
+  []
+  (let [cache-configs     (t2/select :model/CacheConfig :strategy :duration :refresh_automatically true)
+        configs->card-ids (cache-configs->card-ids cache-configs)]
+    (when (seq cache-configs)
+      (let [base-queries          (t2/select :model/Query (duration-queries-to-rerun-honeysql configs->card-ids false))
```

Comment:
nit: Do you think its worth moving `configs->card-ids` into the inner let, so we don't compute it when not `(seq cache-configs)`? I think it'll just be an empty map, but it may make the flow easier to see at a glance.